### ISSUE TITLE
Integer overflow in mean()

### ISSUE THREAD

**Issue by kagalenko-m-b** [JuliaLang/Statistics.jl#22]
```
julia> mean([1,1]*typemax(Int))
-1
julia> versioninfo()
Julia Version 1.3.1
Commit 2d5741174c (2019-12-30 21:36 UTC)
Platform Info:
  OS: Linux (x86_64-pc-linux-gnu)
  CPU: Intel(R) Core(TM) i3 CPU         540  @ 3.07GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-6.0.1 (ORCJIT, westmere)
Environment:
  JULIA_NUM_THREADS = 8
```

**Comment by nalimilan** [comment]
Yes, that's expected, just like `typemax(Int)+typemax(Int)` doesn't give the correct result. See https://docs.julialang.org/en/v1/manual/faq/#faq-integer-arithmetic-1.

**Comment by kagalenko-m-b** [comment]
That would have been expected of the function with an integer result. However, mean()
returns float and having silent integer overflow within is setting up a user for a nasty
unexpected bugs. Perhaps, you could reconsider closing that issue?

**Comment by nalimilan** [comment]
Discussion at https://discourse.julialang.org/t/mean-of-integers-overfloats-bug-or-expected-behaviour/34767/3.

**Comment by stevengj** [comment]
@nalimilan, since this function produces a floating-point result, I don't agree that it should overflow for integer inputs.

(This is analogous to the `hypot` function — as long as the output is representable, we shouldn't overflow on intermediate calculations.)

**Comment by stevengj** [comment]
For example, an implementation like this:
```
function _mean(A::AbstractArray, ::Colon)
    isempty(A) && return sum(A) / length(A) # or maybe: throw(ArgumentError("mean requires non-empty array"))
    n = length(A)
    T = typeof(first(A) / n)
    return sum(x -> convert(T, x), A) / length(A)
end
```

**Comment by stevengj** [comment]
Note that the `mean(A, dims=...)` variant already does the promotion before summing, so it avoids overflow already:
```jl
julia> mean([1,1]*typemax(Int), dims=1)
1-element Array{Float64,1}:
 9.223372036854776e18
```

**Comment by stevengj** [comment]
@kagalenko-m-b, want to take a crack at a pull request?  Just need to use the above implementation for `_mean` and then add a couple of tests.

**Comment by kagalenko-m-b** [comment]
@stevengj, can take a look, what tests you have in mind in addition to the one that is in my original issue?

**Comment by stevengj** [comment]
Tests that come to mind:
```jl
@test mean(fill(typemax(Int), 10)) == mean(float, fill(typemax(Int), 10)) ≈ float(typemax(Int))  # avoid integer overflow (#22)
let x = rand(10000)
    @test mean(x) == sum(x) / length(x) # mean should use sum's accurate pairwise algorithm
end
@test mean(Number[1, 1.5, 2+3im]) == 1.5+1im # mixed-type array
const ≜ = isequal
@test mean(Float64[]) ≜ mean(Int[]) ≜ NaN
```

**Comment by kagalenko-m-b** [comment]
> Note that the `mean(A, dims=...)` variant already does the promotion before summing, so it avoids overflow already:

Wow, different result when supplied with a redundant argument.

It actually **is** a bug.

**Comment by JeffBezanson** [comment]
Triage agrees this should not overflow.

**Comment by stevengj** [comment]
Fixed by #25.
